File: spring-core-lab-solutions/Lab-12-solution/petclinic/src/test/java/com/petclinic/core/VisitServiceUnitTest.java
package com.petclinic.core;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import java.time.LocalDate;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.openMocks;

class VisitServiceUnitTest {
    @Mock
    private VisitRepository visitRepository;
    @InjectMocks
    private VisitService visitService;

    @BeforeEach
    public void setup() {
        openMocks(this);
    }

    @Test
    public void shouldFindVisitByReferenceNumber() {
        var visit = new Visit(0, "V01-23", LocalDate.of(2013, 12, 21), "Teeth whitening");
        when(this.visitRepository.findByReferenceNumber("V01-23")).thenReturn(Optional.of(visit));
        assertThat(this.visitService.findByReferenceNumber("V01-23").getReferenceNumber()).isEqualTo("V01-23");
    }

}

File: spring-core-lab-solutions/Lab-kotlin/src/test/kotlin/com/petclinic/core/VisitServiceUnitTest.kt
package com.petclinic.core

import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import java.time.LocalDate
import java.util.*

internal class VisitServiceUnitTest {
    private val visitRepository = Mockito.mock(VisitRepository::class.java)
    private val visitService = VisitService(visitRepository)

    @BeforeEach
    fun setup() {
        MockitoAnnotations.openMocks(this)
    }

    @Test
    fun shouldFindVisitByReferenceNumber() {
        val visit = Visit(1, LocalDate.now(), "teeth whitening", null, null)
        Mockito.`when`(visitRepository.findById(1)).thenReturn(Optional.of(visit))

        val foundVisit = visitService.findById(1)

        assertThat(foundVisit?.id).isEqualTo(1)
    }
}
